tests/repository/test_repository.py

```diff
@@ -561,21 +586,20 @@ def test_walk(repository, generate_directory):
     ]
 
 
-@pytest.mark.parametrize('path', ('.', 'relative'))
-def test_copy_tree(repository, generate_directory, tmp_path_parametrized, path):
+# yapf: disable
+@pytest.mark.parametrize(('path', 'expected_hierarchy'), (
+    (None, {'file_a': b'a', 'relative': {'file_b': b'b', 'sub': {'file_c': b'c'}}}),
+    ('.', {'file_a': b'a', 'relative': {'file_b': b'b', 'sub': {'file_c': b'c'}}}),
+    ('relative', {'file_b': b'b', 'sub': {'file_c': b'c'}}),
+    ('relative/sub', {'file_c': b'c'}),
```

Comment:
I am confused about this, I think this should be like this? But this is what you going to fix?

```
('relative/sub', {'sub': {'file_c': b'c'}}),
```

I made a test in my command line:

```bash
╭─jyu at theospc38 in /tmp
╰─○ mkdir target_d
╭─jyu at theospc38 in /tmp
╰─○ mkdir -p relative/sub
╭─jyu at theospc38 in /tmp
╰─○ touch relative/b
╭─jyu at theospc38 in /tmp
╰─○ touch relative/sub/c
╭─jyu at theospc38 in /tmp
╰─○ cp -r relative/sub target_d
╭─jyu at theospc38 in /tmp
╰─○ ls target_d 
sub
╭─jyu at theospc38 in /tmp
╰─○ ls target_d/sub 
c
```

My `cp` version is `cp (GNU coreutils) 8.30`.

Reply:
You are right, the behavior of `cp` is different. Note that tat is not just for the last test case but also the penultimate. Essentially, the last directory in the `path` is kept. One would have to do `cp -r relative/sub/* target_d` to achieve what is currently suggested by the test.

I think the docstring of `copytree` is consistent with the implemented behavior. It is just that my commit message regarding the parallel with `cp` is incorrect. I am wondering if that should be the behavior though. I am actually mirroring the behavior of `shutil.copytree`. If you run `shutil.copytree('relative/sub', 'target_d')`, it would simply create `target_d` and copy the file `c` into it. Which is what `Repository.copy_tree` currently does.

Also, if we were to emulate the `cp` behavior, then I am not sure what someone would have to do if they *didn't* want to keep the `sub` directory. In `cp` you would a glob, but that is not supported in the Python API.

So in short, I think we should keep the current implementation but I correct the commit message saying that it mirrors `shutil.copytree` and remove the mention of `cp`. Would that work?

Reply:
Sounds good to me. I'll continue to review along this line. Thanks for the explanation!